05libraries/sec01ChoosingLibraries.md

```diff
@@ -0,0 +1,188 @@
+
+When choosing a library to use with your own project, try to use a *stable* version, i.e. one with a stable interface. 
+
+### Efficiency: Is the library fast enough?
+
+Libraries, particularly the good ones, tend to be well-optimised, that is their algorithms and data structures have been tweaked to get the best performance. For performance-critical libraries (like many used in numerical computations) the library developers should include some details about the performance of the library in its documentation. This is where you should ideally look for information about the performance. Otherwise, try to find comparisons with other, similar libraries to understand the performance.
+
+While you are unlikely to beat a library's performance with a custom algorithm, sometimes custom code can be faster due to a tradeoff between flexibility and performance. If you have already used the library but think you might be able to beat a library's performance:
+
+1. test the performance of the library's implementation
+2. write some unit tests using the library's implementation
+3. write your custom implementation
+4. modify the unit tests to test your implementation
+5. test the performance of your custom code and compare to the library performance
+
+By testing correctness and performance on both the library and your custom code, you can understand whether it's worth it to commit to either.
+
+### Up-to-date: Is the library regularly updated?
+
+Libraries that are not regularly maintained can "rot", that is:
+
+- bugs don't get fixed
+- bugs in dependencies don't get fixed
+- new language features break the library
+- newer, safer language features don't get introduced
+- advances in packaging make it more difficult to install
+
+In general though, we want to avoid these issues, so consider these questions when deciding if the library is suitably up-to-date:
+
+* When was the last release?
+* Is there a sensible versioning scheme (e.g. [semantic versioning][semver])?
+* Is a [changelog][changelog] provided with each new release?
+* Is there a suitable release schedule?
+* Is the code developed on the open (e.g., on GitHub)?
+  * How often are there commits?
+
+You should develop your own intuition for what you consider "suitably up-to-date" but here are some heuristics of mine:
+
+- If a library has been updated within the last year, it's probably good.
+- If a library is very small, it probably doesn't need many updates, so longer releases are fine.
+- If a library is very old (like some numerical libraries) then it is so well-used, there probably aren't many bugs left, so a release over ten years ago is still probably okay (but might not be very efficient on modern hardware).
```

Comment:
Worth pointing out explicitly that you do need to distinguished between well established old libraries and abandoned old libraries!

Reply:
e.g. some knowledge of usage in the field is useful